lib/datadog/tracing/contrib/active_support/cache/instrumentation.rb

```diff
@@ -10,96 +10,63 @@ module Contrib
-                  payload[:action] == Ext::RESOURCE_CACHE_MGET &&
-                  current_span.try(:resource) == Ext::RESOURCE_CACHE_MGET
-                )
-
-              tracing_context = payload.fetch(:tracing_context)
+            # @param action [String] type of cache operation. Will be set as the span resource.
+            # @param key [Object] redis cache key. Used for actions with a single key locator.
+            # @param multi_key [Array<Object>] list of redis cache keys. Used for actions with a multiple key locators.
+            def trace(action, key: nil, multi_key: nil)
+              return yield unless enabled?
 
               # create a new ``Span`` and add it to the tracing context
-              service = Datadog.configuration.tracing[:active_support][:cache_service]
-              type = Ext::SPAN_TYPE_CACHE
-              span = Tracing.trace(Ext::SPAN_CACHE, service: service, span_type: type)
-              span.resource = payload.fetch(:action)
-
-              span.set_tag(Tracing::Metadata::Ext::TAG_COMPONENT, Ext::TAG_COMPONENT)
-              span.set_tag(Tracing::Metadata::Ext::TAG_OPERATION, Ext::TAG_OPERATION_CACHE)
-
-              tracing_context[:dd_cache_span] = span
-            rescue StandardError => e
-              Datadog.logger.debug(e.message)
+              Tracing.trace(
+                Ext::SPAN_CACHE,
+                span_type: Ext::SPAN_TYPE_CACHE,
+                service: Datadog.configuration.tracing[:active_support][:cache_service],
+                resource: action
+              ) do |span|
+                span.set_tag(Tracing::Metadata::Ext::TAG_COMPONENT, Ext::TAG_COMPONENT)
+                span.set_tag(Tracing::Metadata::Ext::TAG_OPERATION, Ext::TAG_OPERATION_CACHE)
+
+                set_backend(span)
+                set_cache_key(span, key, multi_key)
+
+                yield
+              end
             end
 
-            def finish_trace_cache(payload)
-              return unless enabled?
-
-              # retrieve the tracing context and continue the trace
-              tracing_context = payload.fetch(:tracing_context)
-              span = tracing_context[:dd_cache_span]
-              return unless span && !span.finished?
-
-              begin
-                # discard parameters from the cache_store configuration
-                if defined?(::Rails)
-                  store, = *Array.wrap(::Rails.configuration.cache_store).flatten
-                  span.set_tag(Ext::TAG_CACHE_BACKEND, store)
-                end
+            # In most of the cases, `#fetch()` and `#read()` calls are nested.
+            # Instrument both does not add any value.
+            # This method checks if these two operations are nested.
+            def nested_read?
+              current_span = Tracing.active_span
+              current_span && current_span.name == Ext::SPAN_CACHE && current_span.resource == Ext::RESOURCE_CACHE_GET
```

Comment:
I think this condition changed a bit from what it was before refactoring:

```ruby
    return if current_span.try(:name) == Ext::SPAN_CACHE &&
    (
        payload[:action] == Ext::RESOURCE_CACHE_GET &&
        current_span.try(:resource) == Ext::RESOURCE_CACHE_GET ||
        payload[:action] == Ext::RESOURCE_CACHE_MGET &&
        current_span.try(:resource) == Ext::RESOURCE_CACHE_MGET
    )
```

It used `current_span.try(:resource)` - I am not very familiar with codebase so I have no idea whether `current_span` really could have no `resource` method

Reply:
It cannot: `current_span` can be `nil`, but not have missing methods. It's our very own [SpanOperation](https://github.com/DataDog/dd-trace-rb/blob/afc4a025c7c8717050901f1419b7b21856aa9a2a/lib/datadog/tracing/span_operation.rb).